src/steamship/plugin/capabilities.py

```diff
@@ -0,0 +1,291 @@
+    with extra fields appended to them.  This is not included in the init path for a class by default, with the
+    intention that classes define their names at a class-definition-level.
+
+    Make an effort to namespace these by organization, since this is technically extensible.
+    """
+
+    name: str = Field(init_var=False, default=None)
+    """Name of the capability as an instance field.
+
+    The base class of capabilities doesn't have a NAME, but can represent the unresolved Capabilities that clients are
+    requesting.  See CapabilityImpl for the setting of this field.
+    """
+
+    request_level: RequestLevel = RequestLevel.NATIVE
+
+    def is_plugin_support_valid(
+        self, support_level: Optional[RequestLevel]
+    ) -> Tuple[bool, Optional["Capability.Response"]]:
+        """Checks if the plugin fulfills the capability request level for this specific capability.
+
+        Returns a basic Response if the plugin supports the capability and the support level is at least what the
+        consumer asked for.
+
+        :param support_level: The level at which the Plugin asserts that it can support the requested capability
+        :return: a bool that indicates whether the Plugin support level is valid, plus an Optional Response for this
+          capability's support, if applicable.
+        """
+        if self.request_level == RequestLevel.OPTIONAL and support_level is None:
+            # We don't support it, but it's optional.
+            return True, None
+        if support_level is None:
+            # Not optional, but we don't support it.
+            return False, None
+        if self.request_level == RequestLevel.NATIVE and support_level != RequestLevel.NATIVE:
+            # They want native support, but we don't support it at a native level.
+            return False, None
+        # They want NATIVE support, and we support that, OR
+        # They want BEST_EFFORT support, and we support that, OR
+        # They want BEST_EFFORT and we support native
+        return True, Capability.Response(fulfilled_at=support_level)
+
+    class Response(BaseModel, extra=Extra.allow):
+        """Response regarding a specific capability served by the plugin.
+
+        Responses indicate at minimum the level at which they served the request-level.  They may
+        also specify additional metadata.
+        """
+
+        CAPABILITY_NAME: ClassVar[str]
+
+        name: str = Field(init_var=False, default=None)
+        fulfilled_at: RequestLevel
+
+
+class CapabilityImpl(Capability, extra=Extra.forbid):
+    # TODO (PR): The Extra.forbid here is to enforce clamping of deserialization, but it may just be simpler to leave
+    #  that up to individual capabilities?  My goal here is to prevent accidental breakage of contract when someone
+    #  provides specific metadata that makes it so e.g. a plugin with an older view of the world thinks it has Best
+    #  Effort support but can't because of those extra requests.
```

Comment:
Discussion: adding extra details to a capability seems dangerous because it adds extra things to support that a plugin may not know about.

Also note to self, the extra as a metaclass argument here should get moved into a Config inner class for consistency.